File: plugins/_shared/scripts/skill_paths.py
#!/usr/bin/env python3
"""Version-resolved paths to plugin skills in the Codex cache.

The cache nests skills under a version dir (<plugin>/<version>/skills/<skill>/),
so a hardcoded <plugin>/skills/... path does not exist and a deny hint would
point the agent at a missing file. This resolves the latest version segment.
"""
import os

PLUGINS_DIR = os.path.expanduser("~/.codex/plugins/cache/fusengine-codex")


def _semver_key(version: str) -> tuple:
    """Sort key for semantic versions (1.10.0 > 1.2.5, not lexicographic)."""
    return tuple(int(p) if p.isdigit() else 0 for p in version.split("."))


def skill_md(plugin: str, skill: str) -> str:
    """Return the version-resolved path to <plugin>/skills/<skill>/SKILL.md."""
    root = os.path.join(PLUGINS_DIR, plugin)
    rest = os.path.join("skills", skill, "SKILL.md")
    direct = os.path.join(root, rest)
    if os.path.isfile(direct):
        return direct
    try:
        versions = [v for v in os.listdir(root)
                    if os.path.isfile(os.path.join(root, v, rest))]
    except OSError:
        return direct
    return os.path.join(root, max(versions, key=_semver_key), rest) if versions else direct

File: plugins/core-guards/scripts/_shared/skill-paths.ts
/**
 * skill-paths.ts — bundle-safe TS port of _shared/scripts/skill_paths.py.
 *
 * The Codex cache nests skills under a version dir
 * (<plugin>/<version>/skills/<skill>/), so a hardcoded <plugin>/skills/... path
 * does not exist; a deny hint would point at a missing file. Resolves the latest
 * semantic-version segment. Kept dependency-free so the bundler inlines it.
 */
import { homedir } from "node:os";
import { join } from "node:path";
import { existsSync, readdirSync, statSync } from "node:fs";

const PLUGINS_DIR = join(homedir(), ".codex", "plugins", "cache", "fusengine-codex");

/** Semantic-version sort key (1.10.0 > 1.2.5, not lexicographic). */
function semverKey(version: string): number[] {
  return version.split(".").map((p) => (/^\d+$/.test(p) ? Number(p) : 0));
}

/** Compare two semver keys; returns positive when a > b. */
function semverCmp(a: number[], b: number[]): number {
  const n = Math.max(a.length, b.length);
  for (let i = 0; i < n; i++) {
    const d = (a[i] ?? 0) - (b[i] ?? 0);
    if (d !== 0) return d;
  }
  return 0;
}

/**
 * Return the version-resolved path to <plugin>/skills/<skill>/SKILL.md.
 * @param plugin - Plugin name (cache dir under fusengine-codex).
 * @param skill - Skill slug.
 * @returns Absolute SKILL.md path (latest version, or the direct path as fallback).
 */
export function skillMd(plugin: string, skill: string): string {
  const root = join(PLUGINS_DIR, plugin);
  const rest = join("skills", skill, "SKILL.md");
  const direct = join(root, rest);
  if (existsSync(direct)) return direct;
  let versions: string[];
  try {
    versions = readdirSync(root).filter((v) => {
      try {
        return statSync(join(root, v, rest)).isFile();
      } catch {
        return false;
      }
    });
  } catch {
    return direct;
  }
  if (!versions.length) return direct;
  const best = versions.reduce((a, b) => (semverCmp(semverKey(a), semverKey(b)) >= 0 ? a : b));
  return join(root, best, rest);
}
